File: 2026POTD/JANUARY/07-01-2026/1339.go
func maxProduct(root *TreeNode) int {
    const MOD int64 = 1000000007
    var total, best int64

    var sum func(*TreeNode) int64
    sum = func(node *TreeNode) int64 {
        if node == nil {
            return 0
        }
        return int64(node.Val) + sum(node.Left) + sum(node.Right)
    }

    total = sum(root)

    var dfs func(*TreeNode) int64
    dfs = func(node *TreeNode) int64 {
        if node == nil {
            return 0
        }
        s := int64(node.Val) + dfs(node.Left) + dfs(node.Right)
        prod := s * (total - s)
        if prod > best {
            best = prod
        }
        return s
    }

    dfs(root)
    return int(best % MOD)
}

File: 2026POTD/JANUARY/07-01-2026/1339.rb
def max_product(root)
  mod = 1_000_000_007
  @best = 0

  def sum(node)
    return 0 unless node
    node.val + sum(node.left) + sum(node.right)
  end

  total = sum(root)

  def dfs(node, total)
    return 0 unless node
    s = node.val + dfs(node.left, total) + dfs(node.right, total)
    @best = [@best, s * (total - s)].max
    s
  end

  dfs(root, total)
  @best % mod
end